Src/Services.Repo.pas

```diff
@@ -0,0 +1,179 @@
+unit Services.Repo;
+
+interface
+
+uses
+  Services.IRepo,
+
+  Spring.Collections,
+
+  System.SysUtils,
+  System.Variants,
+  System.SyncObjs,
+
+  FireDAC.VCLUI.Wait,
+  FireDAC.Stan.Intf, FireDAC.Stan.Option,
+  FireDAC.Stan.Error, FireDAC.UI.Intf, FireDAC.Stan.Def,
+  FireDAC.Stan.Pool, FireDAC.Stan.Async, FireDAC.Phys,
+  Data.DB, FireDAC.Comp.Client, FireDAC.Phys.SQLite,
+  Firedac.DApt;
+
+type
+
+  TRepo = class(TInterfacedObject, IRepo)
+    private
+      FSQLLite: TFDConnection;
+      FManager: TFDCustomManager;
+      FCT: TCriticalSection;
+
+    public
+      constructor Create;
+
+      function GetFiles(aQuery: string): IList<TSearchResult>;
+      procedure AddFile(aFileName: string; aPath: string);
+      procedure BatchAddFiles(aFiles: IList<string>);
+
+      destructor Destroy; override;
+  end;
+
+
+implementation
+
+uses
+  System.IOUtils;
+
+{ TRepo }
+
+procedure TRepo.AddFile(aFileName, aPath: string);
+begin
+  FCT.Enter;
+  try
+    var lQuery := TFDQuery.Create(nil);
+    lQuery.Connection := FSQLLite;
+    try
+      with lQuery do
+      begin
+        Open('SELECT Path FROM files WHERE Path=?', [aPath]);
+
+        if RecordCount = 0 then
+        begin
+          SQL.Clear;
+          SQL.Add('INSERT INTO files (Path,Name) VALUES (:path,:name)');
+          Params.ParamByName('path').AsString := aPath;
+          Params.ParamByName('name').AsString := aFileName;
+
+          Execute;
+        end;
+      end;
+    finally
+      FreeAndNil(lQuery);
+    end;
+  finally
+    FCT.Leave;
+  end;
+end;
+
+procedure TRepo.BatchAddFiles(aFiles: IList<string>);
+begin
+  FCT.Enter;
+  try
+    var lQuery := TFDQuery.Create(nil);
+    lQuery.Connection := FSQLLite;
+    try
+      with lQuery do
+      begin
+        SQL.Add('delete from files;');
+
+        for var lFile in aFiles do
+        begin
+          var lFileName := TPath.GetFileName(lFile);
+          SQL.Add('INSERT INTO files (Path,Name) VALUES ('''+lFIle+''','''+lFileName+''');');
+        end;
+
+        Execute;
+      end;
+    finally
+      FreeAndNil(lQuery);
+    end;
+  finally
+    FCT.Leave;
+  end;
+end;
+
+constructor TRepo.Create;
+begin
+  FCT := TCriticalSection.Create;
+
+  FManager := FDManager;
+
+  FManager.SilentMode := True;
+
+  FSQLLite := TFDConnection.Create(nil);
+  FSQLLite.DriverName := 'SQLite';
+
+  FSQLLite.Open;
+
+
+  var lQuery := TFDQuery.Create(nil);
+  lQuery.Connection := FSQLLite;
+  try
+    with lQuery do
+    begin
+      SQL.Add('''
+                  CREATE TABLE `files` (
+                    `Path`	TEXT NOT NULL UNIQUE,
+                    `Name`	TEXT NOT NULL,
+                    PRIMARY KEY(Path)
+                  );
```

Comment:
First time I see table name be all lower case but column names be sentence casing

Reply:
Fair point 😄